### ISSUE TITLE
Make possible set the Margin of PhysicsBody2D->Collision as 0.0

### ISSUE THREAD

**Issue by ferreiradaselva** [godotengine/godot#7348]
OS: Linux, x86_64
Version: 2.2 alpha

This is a feature-request (or enhancement). Currently, the minimum value for margin grow is 0.001.

I'm making a "pixel-perfect" demo which I want to make open, and everything works fine, including sticking to slopes. The objective is to make it work like a classic platformer (which is a bit different from the Kinematic Character demo available), and is based on the Game Maker tutorial by Shaun on youtube.

However, if I want  to make a character of, for example, 8x8 pixels to walk between a little space of 8x8 space, I need to set the collision shape of the character to 7x7 (3.5, 3.5 of extents) and margin 0.5.

If I understand right, the margin growing will make the collision shape of the character act as if it was 8x8, and that's why it is currently fitting right.

It would be nice to be able to set the margin as 0.0 and just use the extents, so anyone else can make new collision shapes without having to worry about the sum of extents + margin grow.

Anyway, thanks for the great work on Godot!

**Comment by RebelliousX** [comment]
@felipefs Trust me, I will save you lots of headache and time, what you are trying to achieve is not possible with current physics bodies in Godot.

Even if you are allowed to change the margin to 0.0, it will not work for "pixel-perfect" games. You have to write your own collision detection behavior (with the help of wold 2d direct space state intersection functions of course, otherwise, the code will be lengthy and you are better off using C++ to write the whole game) and decide what to do based on that.

The reason is that current physics bodies if you tell them to move and find themselves stuck in a collision, they will try to free themselves to a non-colliding positions. Those positions will not be integral values, even if you try to round off, floor, ceil, cast to integer, stepify, snap values...etc! You will end up with rounding issues fast due to lost precision.

You might get away with using kinematic body for example if your game is simple enough, that is, all what you need is simple movement and it doesn't require the player for example to push crates, or not allowing pushing 2 stacked crates ...etc! Especially if the game is grid-based, with precise in grid movements is a must.

With current physics bodies, the player can end up standing in mid air, with only 1 pixel of a square's corner touching the player's collision body's corner, making it smaller or removing the margin of collision shape can and will cause imprecise positioning. 

TL;DR: Do collision checking using intersect functions of the wold2d direct space state, and move objects using set_pos().

**Comment by felipefs** [comment]
@RebelliousX The thing is: I have it working... mostly. With exception of that margin grow that is annoying to keep track of every time I need to create a new collision shape.

However, from your comment, I believe I could use only Area2D (instead of static and kinematic bodies) for **every** object and see if it's colliding with each other. In terms of performance, would using Area2D for each object cause a slowdown?

About the "pixel-perfect" term, I used with quotation marks because I know it's not exactly pixel-perfect. Before using Godot I achieved that using my own C engine, and I wasn't using true pixel-perfection either, but the result was the same as we see in GMs tutorial (by Shaun as cited above). So, even with Godot built-in engine, as long as we have a mean to preview collision, i believe it's possible.

Meanwhile, I will try the method above (using Area2D for each object) and I will let you know.

**Comment by RebelliousX** [comment]
Area2D will report collisions once when entering and once when exiting. If you keep that in mind and don't
use forces such as gravity each frame.

For me, I use physics bodies just to make it easier for myself when coding to get the type of colliding objects, static (ground) vs kinematic (enemies or any moving object). But I don't rely on `move()` functions nor physics for platform games.

What I did, and it works perfectly fine for me. 
```
...
world_space = get_world_2d().get_direct_space_state()
.
.
colliders = world_space.intersect_point(Vector2(pos.x + offset, pos.y + offset))
if (not colliders.empty()):
	for collider in colliders:
		if (collider["collider"].is_in_group(Game_Globals.GROUP_SOLIDS) or 
			(collider["collider"].get_parent().has_method("can_stand_on") and 
			collider["collider"].get_parent().can_stand_on())):
.... do whatever you want here...
```

**Comment by felipefs** [comment]
@RebelliousX Thanks, I will try using World2D! :)

**Comment by felipefs** [comment]
Well, as I tried the following solution without success:

1. Work with Area2D, but it can't be tested for overlapped bodies after using ```set_pos()```.
2. Work with Shape2D API, but you can't **preview** overlapping with unknown shapes.

I will keep using the KinematicBody2D, but still with the inconvenience of having to keep track of the sum **extents + margin grow** for each CollisionShape2D created. I'm not a dev, so I don't know if I'm presuming right about the relationship between extens and margin grow. I'm assuming that from what I read in the documentation and tooltip of each property. Correct me If I'm wrong.

I will keep the issue open, unless the devs says otherwise. But I think that could have that option to set margin grow to "0.0".

**Comment by Yesyoor** [comment]
Wow after winding up me head for days through the Godot physics engine I am again frustrated about Godot being not abstract enough. 

I try to make a Tetris clone were the blocks are draggable and collide with new spawning and falling blocks. Since it is Tetris it needs to be pixel perfect.

How can an engine be complicated enough to contain physics in this form but not simple enough to have 0 margin and 0 forces whatsoever upon collisions.

**Comment by Calinou** [comment]
> How can an engine be complicated enough to contain physics in this form but not simple enough to have 0 margin and 0 forces whatsoever upon collisions.

Physics margins are needed for performance reasons. All modern physics engines feature some kind of margins, but you just happen not to notice them as collision shapes' sizes are usually adjusted to compensate for them.

> Area2D is not an option since overlaps may not take place if one was to drag an object too fast (skipping other blocks or wall areas)

You can increase **Physics Fps** in the project settings to reduce tunneling. With relatively simple physics, it should be possible to set it to 120 or even 180 (while also reducing jittering due to stacked RigidBodies).

That said, using a raycast to perform intermediate checks for the dragged object doesn't sound unreasonable to me. It's something that is done fairly often in games in production.

**Comment by Yesyoor** [comment]
> Physics margins are needed for performance reasons. All modern physics engines feature some kind of margins, but you just happen not to notice them as collision shapes' sizes are usually adjusted to compensate for them.

Thanks for replying! I actually solved it yesterday this way, using kinematic body and handling the position myself.

I only use the move_and_collide(v, true,true, test_mode = true) function in test mode to get the collisions and then handle setting the positions myself. So the kinematic body's don't move using the move API but instead (and different than recommended in the docs) I simply set the position by code.

I make sure once I drag something and the dragged object collides with an other object, the dragged won't move at all
( one can use collision.travel to set it aligned to static objects as I did with the borders, you ll notice the difference when I move an object so fast that it would tunnel the other. For simplicity I did it only for the walls ). For other moving object I test if they collide with an object that is beeing dragged, and if so it moves it with it ( Thats why in this case the other object didn't move at all, but I could allign it with the top and sides, since the not dragged object fall down only).

This pretty much solved it, there are some collisions where I'd need to set their position to position = Vector2( floor(position.x) ,floor(position.y)) or to their grid cell ( on mouse released).

If one happens to run into the same issue I can share the code used:

https://user-images.githubusercontent.com/17583446/152781581-33bc0ad4-e529-4a7b-9f43-8de8a5801f75.mov

**Comment by Yesyoor** [comment]
Ok so I tried to go further and failed :D more than a single tile block can't be formed, a cube or a line would all need their separate collsion checks, probably recursive and I won't even try it out since I can imagine the FPS would drop too hard.

> Physics margins are needed for performance reasons. All modern physics engines feature some kind of margins, but you just happen not to notice them as collision shapes' sizes are usually adjusted to compensate for them.

The margin is just the biggest problem, even using move_and_collide() in test mode would not give expected results.
I've build this draggable Tetris from scratch a lot of times now in the past two weeks. Since Tetris is really not a performance heavy thing and I actually don't want to use a kinematic just to have the collide function, dragging properties like materials and what not with me, how would you solve a draggable Tetris where blocks can land on top of the dragged block, pushing it down? Is it really this hard for Godot?


I feel the only way left is implementing my own ray-cast system, or go the recursive way ( 1 collides with 2, check if 2 can be pushed down, if not check if 3 can be pushed down and so on). This would definitely break the object oriented way in forms of organising which tile needs to update and which has already been updated recursive by others.
Also I would like to resume a defined single tile and compose it to a block or line of multiple tiles (like in Tetris) only overriding the drag function (and have references to the group members to drag them along), but collision gets a mess there too, checking the own groups first collision, checking the colliding tile if it is part of a group and so on.

Can someone give me a hint?

**Comment by Yesyoor** [comment]
> That said, using a raycast to perform intermediate checks for the dragged object doesn't sound unreasonable to me. It's something that is done fairly often in games in production.

I found this on reddit, and he seems to use raycasts to but ran into the problem of 2d direct sate not beeing updated when moving an object by setting its position. Is it still a good way to go and is it possible to manipulate the direct state?
